cpp/dolfinx/graph/KaHIP.cpp

```diff
@@ -66,113 +65,111 @@ graph::AdjacencyList<std::int32_t> dolfinx::graph::KaHIP::partition(
+            halo_cell_to_remotes[i] = std::set<std::int32_t>();
+          halo_cell_to_remotes[i].insert(remote);
+        }
       }
     }
-  }
 
-  // Do halo exchange of cell partition data
-  std::vector<std::vector<std::int64_t>> send_cell_partition(num_processes);
-  std::vector<std::int64_t> recv_cell_partition;
-  for (const auto& hcell : halo_cell_to_remotes)
-  {
-    for (auto proc : hcell.second)
+    // Do halo exchange of cell partition data
+    std::vector<std::vector<std::int64_t>> send_cell_partition(num_processes);
+    std::vector<std::int64_t> recv_cell_partition;
+    for (const auto& hcell : halo_cell_to_remotes)
     {
-      assert(proc < num_processes);
+      for (auto proc : hcell.second)
+      {
+        assert(proc < num_processes);
 
-      // global cell number
-      send_cell_partition[proc].push_back(hcell.first + elm_begin);
+        // global cell number
+        send_cell_partition[proc].push_back(hcell.first + elm_begin);
 
-      // partitioning
-      send_cell_partition[proc].push_back(part[hcell.first]);
+        // partitioning
+        send_cell_partition[proc].push_back(part[hcell.first]);
+      }
     }
-  }
-
-  // Actual halo exchange
-  dolfinx::MPI::all_to_all(mpi_comm, send_cell_partition, recv_cell_partition);
 
-  // Construct a map from all currently foreign cells to their new
-  // partition number
-  std::map<std::int64_t, std::int32_t> cell_ownership;
-  for (auto p = recv_cell_partition.begin(); p != recv_cell_partition.end();
-       p += 2)
-  {
-    cell_ownership[*p] = *(p + 1);
-  }
+    // Actual halo exchange
+    dolfinx::MPI::all_to_all(mpi_comm, send_cell_partition,
+                             recv_cell_partition);
 
-  // Generate mapping for where new boundary cells need to be sent
-  for (std::int32_t i = 0; i < ncells; i++)
-  {
-    const std::size_t proc_this = part[i];
-    for (std::int32_t j = xadj[i]; j < xadj[i + 1]; ++j)
+    // Construct a map from all currently foreign cells to their new
+    // partition number
+    std::map<std::int64_t, std::int32_t> cell_ownership;
+    for (auto p = recv_cell_partition.begin(); p != recv_cell_partition.end();
+         p += 2)
     {
-      const unsigned long long other_cell = adjncy[j];
-      std::size_t proc_other;
-
-      if (other_cell < elm_begin || other_cell >= elm_end)
-      { // remote cell - should be in map
-        const auto find_other_proc = cell_ownership.find(other_cell);
-        assert(find_other_proc != cell_ownership.end());
-        proc_other = find_other_proc->second;
-      }
-      else
-        proc_other = part[other_cell - elm_begin];
+      cell_ownership[*p] = *(p + 1);
+    }
 
-      if (proc_this != proc_other)
+    // Generate mapping for where new boundary cells need to be sent
+    for (std::int32_t i = 0; i < ncells; i++)
+    {
+      const std::int32_t proc_this = part[i];
+      for (std::int32_t j = xadj[i]; j < xadj[i + 1]; ++j)
```

Comment:
Is `xadj` a raw pointer? Can we use it via Eigen?

Reply:
Yes, there is no need to use it raw here.

Reply:
Using `Eigen::Map` now.